crates/sage/src/enzyme.rs

```diff
@@ -141,75 +161,135 @@ impl Enzyme {
                     continue;
                 }
             }
-            ranges.push(left..right);
+            sites.push(DigestSite{
+                site: left..right,
+                missed_cleavages: 0,
+            });
             left = right;
         }
-        ranges.push(left..sequence.len());
-        ranges
+        sites.push(DigestSite{
+            site:left..sequence.len(),
+            missed_cleavages: 0,
+        });
+        sites
     }
 }
 
 impl EnzymeParameters {
-    fn cleavage_sites(&self, sequence: &str) -> Vec<std::ops::Range<usize>> {
+    fn cleavage_sites(&self, sequence: &str) -> Vec<DigestSite> {
         match &self.enyzme {
             Some(enzyme) => enzyme.cleavage_sites(sequence),
             None => {
                 // Perform a non-specific digest
                 let mut v = Vec::new();
                 for len in self.min_len..=self.max_len {
                     for i in 0..=sequence.len().saturating_sub(len) {
-                        v.push(i..i + len)
+                        v.push(DigestSite{
+                            site: i..i + len,
+                            missed_cleavages: 0}
+                        );
                     }
                 }
                 v
             }
         }
     }
 
+    fn missed_cleavage_sites(&self, sites: &mut Vec<DigestSite>, missed_cleavages: &u8) -> Vec<DigestSite> {
+        let mut missed_cleavage_sites = Vec::new();
+        for cleavage in 1..=(1 + missed_cleavages) {
+            // Generate missed cleavages
+            for win in sites.windows(cleavage as usize) {
+                let start = win[0].site.start;
+                let end = win[cleavage as usize - 1].site.end;
+                missed_cleavage_sites.push(DigestSite{
+                    site: start..end,
+                    missed_cleavages: cleavage - 1,
+                });
+            }
+        }
+        sites.append(&mut missed_cleavage_sites);
+        sites.to_vec()
+    }
+
+    fn is_semi_enzymatic(&self) -> bool {
+        match &self.enyzme {
+            Some(enzyme) => enzyme.is_semi_enzymatic,
+            None => false,
+        }
+    }
+
+    fn semi_enzymatic_sites(&self, sites: &mut Vec<DigestSite>) -> Vec<DigestSite> {
+        let mut semi_enzymatic_sites = Vec::new();
+        for site in sites.iter_mut() {
+            let start = site.site.start;
+            let end = site.site.end;
+            for cut_site in start..end {
+                semi_enzymatic_sites.push(DigestSite{
+                    site: start..cut_site,
+                    missed_cleavages: site.missed_cleavages
```

Comment:
You are correct. Currently, this section of code relies on `seen` seeing the non-missed cleavage semi-enzymatic digested peptides first (with the correct number of `missed_cleavages`) followed by those from peptides with missed cleavages. This preserves the correct number of `missed_cleavages` for all peptides since duplicate peptides are formed but are likely to be seen from "parent" peptides with the lowest possible `missed_cleavages` number first; however, it is not as robust as I would like it to be. 

One optimization is to have DigestSite keep track of the location of missed cleaves and adjust the semi-enzymatic peptide's number accordingly. What do you think?

Reply:
Gotcha, that makes sense, and checks out with the tests too. I think we can probably leave as-is for now - I'd like to refactor digest/peptide generation at some point anyway